### DISCUSSION TITLE
Instrument Channel implementation

### DISCUSSION THREAD

**Discussion by PowiHugh** [pymeasure/pymeasure#648] [Q&A]
Hello,

My last question was a while ago; I have spent a lot more time with Python and Pymeasure since then.

Ultimately that previous question was looking for a general solution to the question, how do I create a pymeasure control for the following:

app.Acquisition.C1.AverageSweeps = 25

Where C1 should be variable (C1-C4,Z1-Z4,M1-M4, etc) and 25 should be the "value" within the control (range 1 to 1000000, already defined as a strict_range)?

I see here: https://github.com/pymeasure/pymeasure/discussions/617

That the concept of restructuring the Instrument class, or extending it, to also describe instrument channels has been proposed and accepted as a good idea.

Is there a branch already which has this implemented?

I'm still working through the variety of tools that Python brings to the table normally but I notice that most of the functions within pymeasure classes contains **kwargs, which is in effect var-args but more flexible.  I'm still reading about these.  Is the presence of this in the Instrument class a way by which I as a humble instrument adder may increase the number of interpreted variables in the get_command and set_command strings?

Thanks in advance,
-Hugh

**Comment by msmttchr** [comment]
> Hello,
> 
> My last question was a while ago; I have spent a lot more time with Python and Pymeasure since then.
> 
> Ultimately that previous question was looking for a general solution to the question, how do I create a pymeasure control for the following:
> 
> app.Acquisition.C1.AverageSweeps = 25
> 
> Where C1 should be variable (C1-C4,Z1-Z4,M1-M4, etc) and 25 should be the "value" within the control (range 1 to 1000000, already defined as a strict_range)?

I would assume that C1-C4,Z1-Z4,M1-M4, etc are sort of instrument channels that need to be configured. Normally an instrument channel is an instrument by itself which is repeated a certain number of times in the enclosing instrument.

An example of how this can be implemented is in https://github.com/pymeasure/pymeasure/commits/master/pymeasure/instruments/keithley/keithley2306.py (thanks to @mfikes)

> 
> I see here: #617
> 
> That the concept of restructuring the Instrument class, or extending it, to also describe instrument channels has been proposed and accepted as a good idea.
> 
> Is there a branch already which has this implemented?
> 

Unfortunately not, due to lack of time mainly.

> I'm still working through the variety of tools that Python brings to the table normally but I notice that most of the functions within pymeasure classes contains **kwargs, which is in effect var-args but more flexible. I'm still reading about these. Is the presence of this in the Instrument class a way by which I as a humble instrument adder may increase the number of interpreted variables in the get_command and set_command strings?

I would say that, if you have a property `pr` with `set_command` set to `":INST:COMMAND %d %d"`, you could then do this:
`inst.pr = (10,20)
`
So you should be able to pass two parameters to `set_command`

I do not think you can pass parameters to `get_command`, though.

Finally, I do not think that **kwargs can  be used for your scope.
If you really need more parameters, you may want to implement a method instead.
> 
> Thanks in advance, -Hugh
